src/prefect/utilities/executors.py

```diff
@@ -39,52 +51,94 @@ def run_with_heartbeat(
 
     @wraps(runner_method)
     def inner(
-        self: "prefect.engine.runner.Runner", *args: Any, **kwargs: Any
+        self: "prefect.engine.cloud.CloudFlowRunner", *args: Any, **kwargs: Any
     ) -> "prefect.engine.state.State":
+
         try:
-            p = None
-            try:
-                if self._heartbeat():
-                    # we use Popen + a prefect CLI for a few reasons:
-                    # - using threads would interfere with the task; for example, a task
-                    #   which does not release the GIL would prevent the heartbeat thread from
-                    #   firing
-                    # - using multiprocessing.Process would release the GIL but a subprocess
-                    #   cannot be spawned from a daemonic subprocess, and Dask sometimes will
-                    #   submit tasks to run within daemonic subprocesses
-                    current_env = dict(os.environ).copy()
-                    auth_token = prefect.context.config.cloud.get("auth_token")
-                    api_key = prefect.context.config.cloud.get("api_key")
-                    tenant_id = prefect.context.config.cloud.get("tenant_id")
-                    api_url = prefect.context.config.cloud.get("api")
-                    current_env.setdefault("PREFECT__CLOUD__AUTH_TOKEN", auth_token)
-                    current_env.setdefault("PREFECT__CLOUD__API_KEY", api_key)
-                    current_env.setdefault("PREFECT__CLOUD__TENANT_ID", tenant_id)
-                    current_env.setdefault("PREFECT__CLOUD__API", api_url)
-                    clean_env = {k: v for k, v in current_env.items() if v is not None}
-                    p = subprocess.Popen(
-                        self.heartbeat_cmd,
-                        env=clean_env,
-                        stdout=subprocess.DEVNULL,
-                        stderr=subprocess.DEVNULL,
-                    )
-            except Exception:
-                self.logger.exception(
-                    "Heartbeat failed to start.  This could result in a zombie run."
-                )
+            use_heartbeat = self._heartbeat()
+        except Exception:
+            use_heartbeat = False
+            logger = get_logger()
+            logger.exception(
+                "Heartbeat process is misconfigured.  This could result in a zombie run."
+            )
+
+        if not use_heartbeat or (prefect.context.config.heartbeat_mode == "off"):
+            configured_heartbeat = no_heartbeat()
+        elif prefect.context.config.heartbeat_mode == "thread":
+            configured_heartbeat = threaded_heartbeat(self.flow_run_id)
```

Comment:
Ah I see, is this why you set the run ID as an attribute?

Reply:
yeah, this is the analogy to `self.heartbeat_cmd`, I'd like to remove at least one of them but I want to get input on how deep we want to go in ripping out old idiom now that heartbeats are cloud-flow-run only, as I mentioned above

Reply:
Yea I see how this could get complicated; ultimately `flow_run_id` _is_ provided in context so you could extract it from there in this line, but I'm not sure that's any better than the mild class mutation you already have